core/arns/src/main/java/software/amazon/awssdk/arns/ArnResource.java

```diff
@@ -0,0 +1,237 @@
+    }
+
+    /**
+     * @return the optional resource qualifier
+     */
+    public Optional<String> qualifier() {
+        return Optional.ofNullable(qualifier);
+    }
+
+    /**
+     * @return a builder for {@link ArnResource}.
+     */
+    public static Builder builder() {
+        return new DefaultBuilder();
+    }
+
+    /**
+     * Parses a string containing either a resource, resource type and resource or
+     * resource type, resource and qualifier into an {@link ArnResource}.
+     *
+     * <p>
+     * Supports fields separated by either ":" or "/".
+     *
+     * <p>
+     * For legacy AWS Arns not following the resourceType:resource:qualifier pattern,
+     * the qualifier field will contain everything after the first two sections separated
+     * by either ":" or "/".
+     *
+     * @param resource - The resource string to parse.
+     * @return {@link ArnResource}
+     */
+    public static ArnResource fromString(String resource) {
+        String splitter = resource.contains("/") ? "/" : ":";
+
+        int resourceTypeColonIndex = resource.indexOf(splitter);
+
+        if (resourceTypeColonIndex < 0) {
+            return ArnResource.builder().resource(resource).build();
+        }
+
+        ArnResource.Builder builder = ArnResource.builder()
+                                                 .resourceType(resource.substring(0, resourceTypeColonIndex));
+
+        int resourceColonIndex = resource.indexOf(splitter, resourceTypeColonIndex);
+        int qualifierColonIndex = resource.indexOf(splitter, resourceColonIndex + 1);
+
+        if (qualifierColonIndex < 0) {
+            builder.resource(resource.substring(resourceTypeColonIndex + 1));
+        } else {
+            builder.resource(resource.substring(resourceTypeColonIndex + 1, qualifierColonIndex));
+            builder.qualifier(resource.substring(qualifierColonIndex + 1));
+        }
+
+        return builder.build();
+    }
+
+    @Override
+    public String toString() {
+        return this.resourceType
+               + ":"
+               + this.resource
+               + ":"
+               + this.qualifier;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        ArnResource that = (ArnResource) o;
+
+        if (!Objects.equals(resourceType, that.resourceType)) {
+            return false;
+        }
+        if (!Objects.equals(resource, that.resource)) {
+            return false;
+        }
+        return Objects.equals(qualifier, that.qualifier);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = resourceType != null ? resourceType.hashCode() : 0;
+        result = 31 * result + (resource != null ? resource.hashCode() : 0);
+        result = 31 * result + (qualifier != null ? qualifier.hashCode() : 0);
+        return result;
+    }
```

Comment:
I don't see test coverage for this code.

Reply:
added tests